tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla

```diff
@@ -0,0 +1,384 @@
+
+Alias == [
+    \* constants
+    BuffSz            |-> BuffSz,
+    MaxOffset         |-> MaxOffset,
+
+    \* regular vars
+    dirty             |-> dirty,
+    length            |-> length,
+    curr              |-> curr,
+    lo                |-> lo,
+    buff              |-> buff,
+    diskPos           |-> diskPos,
+    file_content      |-> file_content,
+    file_pointer      |-> file_pointer,
+
+    \* abstract vars
+    abstract_contents |-> LogicalFileContent]
+
+===============================================================================
+
+--------------------------- MODULE RandomAccessFile ---------------------------
+
+EXTENDS Naturals, Sequences, Common
+
+VARIABLES
+    file_content,
+    file_pointer
+
+vars == <<file_content, file_pointer>>
+
+TypeOK ==
+    /\ file_content \in ArrayOfAnyLength(Symbols \union {ArbitrarySymbol})
+    /\ ArrayLen(file_content) <= MaxOffset
+    /\ file_pointer \in Offset
+
+Init ==
+    /\ file_content = EmptyArray
+    /\ file_pointer = 0
+
+Seek(new_offset) ==
+    /\ new_offset \in Offset
+    /\ file_pointer' = new_offset
+    /\ UNCHANGED <<file_content>>
+
+SetLength(new_length) ==
+    /\ file_content' = TruncateOrExtendFile(file_content, new_length)
+    /\ file_pointer' = Min(file_pointer, new_length)
+
+Read(output) ==
+    /\ output = ArraySlice(file_content, file_pointer, Min(file_pointer + ArrayLen(output), ArrayLen(file_content)))
+    /\ file_pointer' = file_pointer + ArrayLen(output)
+    /\ UNCHANGED <<file_content>>
+
+Write(data) ==
+    /\ file_pointer + ArrayLen(data) <= MaxOffset
+    /\ file_content' = WriteToFile(file_content, file_pointer, data)
+    /\ file_pointer' = file_pointer + ArrayLen(data)
+
+Next ==
+    \/ \E offset \in Offset:
+        \/ Seek(offset)
+        \/ SetLength(offset)
+    \/ \E len \in 1..MaxOffset: \E data \in Array(Symbols, len):
+        \/ Write(data)
+        \/ Read(data)
+
+Spec == Init /\ [][Next]_vars
+
+===============================================================================
+
+-------------------------------- MODULE Common --------------------------------
+
+EXTENDS Naturals, Sequences
+
+CONSTANTS
+    Symbols,
+    ArbitrarySymbol,
+    MaxOffset
+
+Offset == 0..MaxOffset
+
+Min(a, b) == IF a <= b THEN a ELSE b
+Max(a, b) == IF a <= b THEN b ELSE a
+
+\* 0-indexed arrays
+ArrayOfAnyLength(T) == [elems: Seq(T)] \* could use Seq(T), but we don't want a[i] to work; have to use ArrayGet
+Array(T, len) == [elems: [1..len -> T]]
+MkArray(len, f) == [elems |-> [i \in 1..len |-> f[i - 1]]]
+EmptyArray == [elems |-> <<>>]
+ArrayLen(a) == Len(a.elems)
+ArrayToSeq(a) == a.elems
+SeqToArray(seq) == [elems |-> seq]
+ArrayGet(a, i) == a.elems[i+1]
+ArraySet(a, i, x) == [a EXCEPT !.elems[i+1] = x]
+ArraySlice(a, startInclusive, endExclusive) == [elems |-> SubSeq(a.elems, startInclusive + 1, endExclusive)]
+ArrayConcat(a1, a2) == [elems |-> a1.elems \o a2.elems]
+
+WriteToFile(file, offset, data_to_write) ==
```

Comment:
It is interesting to me that most of these functions are about Arrays.  Even `WriteToFile` and `ExtendFile` are relatively easy make generic array operations, maybe `OverlapAt` and `TruncateOrExtendWith`, where both take a placeholder as an argument.  It may simply be worth making this an `Array` module, and then `Common` stays more focused on the spec itself `WriteToFile(file, offset, data_to_write) == Array.Overlap(file, data_to_write, offset, ArbitarySymbol)` (I have no idea why I felt compelled to reverse arguments).

I suppose, otherwise I was going to suggest that we add brief documentation on what the semantics of `WriteToFile` and `ExtendFile` were, but in many ways something like `ExtendFile == Array.TurncateOrExtendWith(file, new_length, ArbitrarySymbol)` does this for you.

Reply:
I am probably overthinking these utility functions, but it struck me that maybe the most intuitive implementation involves slice & concat.

```
TruncateOrExtend(a, T, len) == IF len > ArrayLen(a) THEN ArrayConcat(a, Array(T, len - ArrayLen(a))) ELSE ArraySlice(a, 0, len)

OverlayAt(a1, a2, T, i) == ArrayConcat(TruncateOrExtend(a2, T, i), a2)
```

Reply:
Should any of those utility operators be contributed as an Array module to our CommunityModules?

Reply:
I added some simple documentation to these operators.

> but it struck me that maybe the most intuitive implementation involves slice & concat

I agree that the current definitions are a little ugly. I wrote them after one or two failed attempts to come up with simpler definitions.

The ones you've given are certainly cleaner than my early attempts, and I will adopt your definition for `TruncateOrExtend`. But I might keep the definition I wrote for `WriteToFile`. The equivalent definition in terms of other operators is actually fairly cryptic (your definition for `OverlayAt` needs to include the portion of the contents after the written data if the contents were not written at the end of the array):

```tla
WriteToFile(file, offset, data_to_write) ==
    ArrayConcat(ArrayConcat(
        TruncateOrExtendFile(file, offset),
        data_to_write),
        ArraySlice(file, offset + ArrayLen(data_to_write), ArrayLen(file)))
```

> Should any of those utility operators be contributed as an Array module to our CommunityModules?

Yes! I'll look into doing that soon.

Reply:
> your definition for OverlayAt needs to include the portion of the contents after the written data if the contents were not written at the end of the array

Whoops!  Easy to make things simpler if you forget the critical parts :laughing: